LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/MessageProcessor.cs

```diff
@@ -509,9 +509,17 @@ async Task<DownlinkPktFwdMessage> ProcessDataMessageAsync(LoRaTools.LoRaPhysical
             }
             else
             {
-                datr = this.loraRegion.GetDownstreamDR(rxpk);
-                freq = this.loraRegion.GetDownstreamChannel(rxpk);
-                tmst = rxpk.Tmst + timeWatcher.GetReceiveWindow1Delay(loRaDevice) * 1000000;
+                try
+                {
+                    datr = this.loraRegion.GetDownstreamDR(rxpk);
+                    freq = this.loraRegion.GetDownstreamChannelFrequency(rxpk);
+                    tmst = rxpk.Tmst + this.loraRegion.Receive_delay1 * 1000000;
+                }
+                catch (Exception ex)
+                {
+                    Logger.Log(loRaDevice.DevEUI, ex.Message, LogLevel.Error);
```

Comment:
Wouldn't it make sense to throw a `InvalidDataRateException` inside the Region.GetDownstreamDR() method? Then change the catch to only catch this kind of exception.

Additional, you could add more contextual information when logging:
`Logger.Log(devEUI, $"Error resolving downstream data rate. {ex.Message}", LogLevel.Error);`

Reply:
I am working on this, I created a RegionMappingException to handle DR, but also frequency and other issues related to regions.